src/execution/interpreter_loop.rs

```diff
@@ -1403,6 +1907,367 @@ pub(super) fn run<H: HookSet>(
+                    }
+                    MEMORY_INIT => {
+                        let data_idx = wasm.read_var_u32().unwrap_validated() as DataIdx;
+                        let data = unsafe { store.data.get_unchecked_mut(data_idx) };
+                        let mem_idx = 
+                        // if multi_memory_is_enabled {
+                        //     wasm.read_var_u32().unwrap_validated() as usize
+                        // } else {
+                            wasm.read_u8().unwrap_validated() as usize
+                        // }
+                        ;
+                        let mem = store.mems.get(mem_idx).unwrap_validated();
+                        let mut n: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        let mut s: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        let mut d: i32 = stack.pop_value(ValType::NumType(NumType::I32)).into();
+                        if n < 0 || s < 0 || d < 0 {
+                            return Err(RuntimeError::MemoryAccessOutOfBounds);
+                        }
+
+                        if s as usize + n as usize > data.init.len()
+                            || d as usize + n as usize > mem.data.len()
+                        {
+                            return Err(RuntimeError::MemoryAccessOutOfBounds);
+                        }
+
+                        while n != 0 {
+                            let b = data.init[s as usize];
+                            {
+                                let memarg = MemArg {
+                                    align: 0,
+                                    offset: 0,
+                                };
+
+                                let data_to_store: i32 = b as i32;
+                                let relative_address: u32 = d as u32;
+
+                                let mem = store.mems.get_mut(0).unwrap_validated();
+
+                                // The spec states that this should be a 33 bit integer
+                                // See: https://webassembly.github.io/spec/core/syntax/instructions.html#memory-instructions
+                                // ea => effective address
+                                let ea = memarg.offset.checked_add(relative_address);
+                                let memory_location = ea
+                                    .and_then(|address| {
+                                        let address = address as usize;
+                                        mem.data.get_mut(address..(address + 1))
+                                    })
+                                    .ok_or(RuntimeError::MemoryAccessOutOfBounds)?;
+
+                                memory_location.copy_from_slice(&data_to_store.to_le_bytes()[0..1]);
+                            }
+                            d += 1;
+                            s += 1;
+                            n -= 1;
+                        }
```

Comment:
Please link to the algorithm in the spec (and maybe rename the single charachter variables to something more readable)

Reply:
Linked algorithms. I am opposed to changing variable names, as it is much easier to check the spec implementation this way if you want to. I am open to hearing others have to say about this.
I have also added their meaning at the start of each instruction.